File: microservices/company/src/main/java/com/illunex/invest/company/controller/ProductControllerImpl.java
package com.illunex.invest.company.controller;

import com.illunex.invest.api.common.response.ResponseData;
import com.illunex.invest.api.core.company.controller.ProductController;
import com.illunex.invest.api.core.company.dto.ProductDTO;
import com.illunex.invest.company.service.ProductService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class ProductControllerImpl implements ProductController {
    @Qualifier(value = "CompanyProductService")
    private final ProductService productService;

    @Override
    public ResponseEntity<List<ProductDTO>> getProductList(Long companyId) {
        return ResponseEntity.ok(productService.getProductByCompanyIdx(companyId));
    }

    @Override
    public ResponseEntity<ProductDTO> getMainProduct(Long companyId) {
        return ResponseEntity.ok(productService.getMainProductByCompanyIdx(companyId));
    }

    @Override
    public ResponseEntity<Page<ProductDTO>> getProductPage(Long companyId) {
        return new ResponseEntity<>(productService.getProductPageByCompanyIdx(companyId), HttpStatus.OK);
    }

    @Override
    public ResponseEntity<ResponseData> addProduct(ProductDTO productDTO) {
        return ResponseEntity.ok(ResponseData.builder()
                .errorCode(0)
                .message("success")
                .data(productService.edit(productDTO))
                .build());
    }

    @Override
    public ResponseEntity<ProductDTO> readProduct(Long productId) {
        return ResponseEntity.ok(productService.read(productId));
    }

    @Override
    public ResponseEntity<ResponseData> deleteProduct(Long productId) {
        return null;
    }
}

File: microservices/shop/src/main/kotlin/com/illunex/invest/shop/controller/ProductControllerImpl.kt
package com.illunex.invest.shop.controller

import com.illunex.invest.api.common.response.ResponseData
import com.illunex.invest.api.common.response.ResponseList
import com.illunex.invest.api.core.shop.controller.ProductController
import com.illunex.invest.api.core.shop.dto.PluginDTO
import com.illunex.invest.api.core.shop.dto.ProductDTO
import com.illunex.invest.api.core.shop.dto.PurchaseDTO
import com.illunex.invest.api.core.shop.request.PurchaseRequest
import com.illunex.invest.shop.service.ShopProductService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController
import java.util.stream.Collectors

@RestController
class ProductControllerImpl: ProductController {
    @Autowired var shopProductService:ShopProductService? = null

    override fun getPlugins(companyIdx: Long?): ResponseEntity<ResponseList<ProductDTO>> {
        return ResponseEntity.ok(ResponseList(0, "success", shopProductService!!.getPluginList()))
    }

    override fun getPlugins(ids: List<String>): ResponseEntity<ResponseList<PluginDTO>> {
        return ResponseEntity.ok(ResponseList(0, "success",
        shopProductService!!.getPluginByPluginIds(ids.stream().map { s: String -> s.toLong() }.collect(Collectors.toList()))))
    }

    override fun purchase(purchaseRequest: PurchaseRequest?): ResponseEntity<ResponseData<PurchaseDTO>> {
        var purchaseDTO:PurchaseDTO? = shopProductService!!.purchase(purchaseRequest!!.userIdx, purchaseRequest!!.buyProducts)
        return ResponseEntity.ok(ResponseData(0, "success", purchaseDTO!!))
    }
}
/*
return  ResponseEntity.ok(ResponseList(0,
                "success",
                shopPluginService!!
                        .getPluginByPluginIds(ids
                                .stream()
                                .map {
                                    s: String -> s.toLong()
                                }
                                .collect(Collectors.toList())
                        )
        ))
*/